Support for intervention/image:^3.4.0

kevindierkx opened the pull request:
This PR's primary focus is to support `intervention/image:^3.4.0`, it comes with some breaking changes however.

## Text wrapping

Wrapping of long sentences is now handled by `intervention/image` directly, truncating and the removal of lines of text that wouldn't fit in the containing 'box' is handled by `the-og`.

This heavily relies on the `FontProcessor` implementation of the used driver, eg. Imagick or Gd. Since we need to calculate the dimensions of a `TextBox`, to correctly calculate the 'anchor' point/position of related items, the 'canvas' ie. the `ImageInterface` has been moved from the feature's `render` method to an earlier stage using the added `setCanvas()` method. This is currently placed in the `addFeature()` method, called during the `render` cycle of the layout.

This modifies the signature of the `render()` method and will therefore break custom "Features".

## Addition of the `dimensions()` method

The `dimensions()` method essentially replaces the `getPrerenderedBox()` and `getRenderedBox()` methods. The latter were primarily used to calculate anchors and line lengths. Unless a custom "Feature" would call these directly this shouldn't be as impactful.

## Removal of the `TextModifier` implementation

This class was added at some point to calculate the bounding box of a text block. Since this is now directly handled in either the `dimenstions()` method or the `TextBox` implementation this class is now obsolete.

## Driver agnostic improvements

Because we now directly interact with the image instance for various operations that would require a driver we don't need to initialise drivers or driver related instances ourself. Except for the `PictureBox` everything should now also work with GD, not sure if anyone was waiting for GD support... but hey it's there.

## CTA placement fixes

Some default layouts were missing the preferred position, these have been added and the CTA element is now correctly displayed for all layouts.

## Example output

Using the example from the docs, the image below is now the resulting output.

![image](https://github.com/user-attachments/assets/0634d015-8bdc-4840-88b3-2e033645caea)

simonhamp reviewed (commented):
@kevindierkx this looks great but the tests are failing... I think you'll need to refresh the snapshots and commit those changes

kevindierkx commented:
@simonhamp I've update the snapshots. The placement of the first two using the avatar seems a little off:

- https://github.com/simonhamp/the-og/pull/33/commits/0a2f319450f31cd69f366574046dc4233961ad28#diff-99cce31f1e3bc63184123292bfbf98a8de13f439d2ea0bc3d71e09b18d9580b8
- https://github.com/simonhamp/the-og/pull/33/commits/0a2f319450f31cd69f366574046dc4233961ad28#diff-a129afbd5fdf8d345105e84e8eb853f2e2ac94390c2862ff3f081efa7a952eb9

The other snapshots word-wrap a little differently now when using the intevention wrapping logic, which would be expected.

I'll try to review those avatar layouts next week.

simonhamp commented:
@kevindierkx and [this one](https://github.com/simonhamp/the-og/blob/0a2f319450f31cd69f366574046dc4233961ad28/tests/Integration/__snapshots__/ImageTest__test_basic_image%20with%20data%20set%20basic%20with%20background%20url__1.png) also has something up with the background repeating where it didn't repeat before

kevindierkx commented:
@simonhamp The `PictureBox` element didn't use the new `dimensions()` method and would still assume the base 'box' dimensions when calculating its position. This has been updated to calculate the relative dimensions based on the placement of the picture within the box. See: https://github.com/simonhamp/the-og/pull/33/commits/3c342266cb7fa67b0ef9860a1b263d5031adb27f#diff-d43c718413cd3a9519197f1b78ce3acac5c7ec0c2417190d79fe2f4a97466c42R116

I removed the previous set of updated snapshots and generated a new set. The avatar based layouts aren't even included in the update since they're exactly the same now.

As mentioned in https://github.com/simonhamp/the-og/pull/34#issuecomment-2715185848 the background placement on some of the snapshots would relate to the `Cover` variant, however the default placement behaviour of the background is `Repeat`. Since the tests explicitly compose 'repeating' backgrounds I'm assuming the snapshots were outdated.

All other updated snaphots included in https://github.com/simonhamp/the-og/pull/33/commits/a5a1598c9688c15b00622da96ffd57155fd422a6 now slightly differ due to the Intervention's word-wrapping, which would be expected.

Edit: Rebased with main and applied float to the `size()` method of the `TextBox` similar to #34

simonhamp reviewed (approved):
Great work @kevindierkx 🙌🏼
